jwst/stpipe/utilities.py

```diff
@@ -144,3 +149,63 @@ def folder_traverse(folder_path, basename_regex='.+', path_exclude_regex='^$'):
         for file in files:
             if basename_regex.match(file):
                 yield os.path.join(root, file)
+
+
+def record_step_status(datamodel, cal_step, success=True):
+    """Record whether or not a step completed in meta.cal_step
+
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to modify in place
+
+    cal_step : str
+        The attribute in meta.cal_step for recording the status of the step
+
+    success : bool
+        If True, then 'COMPLETE' is recorded.  If False, then 'SKIPPED'
+    """
+    if success:
+        status = COMPLETE
+    else:
+        status = SKIPPED
+
+    if isinstance(datamodel, Sequence):
+        for model in datamodel:
+            model.meta.cal_step._instance[cal_step] = status
+    else:
+        datamodel.meta.cal_step._instance[cal_step] = status
+
+    # TODO: standardize cal_step naming to point to the official step name
+
+
+def query_step_status(datamodel, cal_step):
+    """Query the status of a step in meta.cal_step
+    
+    Parameters
+    ----------
+    datamodel : `~jwst.datamodels.JwstDataModel` instance
+        This is the datamodel or container of datamodels to check
+        
+    cal_step : str
+        The attribute in meta.cal_step to check
+    
+    Returns
+    -------
+    status : str
+        The status of the step in meta.cal_step, typically 'COMPLETE' or 'SKIPPED'
+    """
+    if isinstance(datamodel, Sequence):
+        status = np.array([getattr(model.meta.cal_step, cal_step, NOT_SET) for model in datamodel])
+        if np.any(status == COMPLETE):
+            return COMPLETE
+        elif np.all(status == NOT_SET):
+            return NOT_SET
+        elif np.all(status == SKIPPED):
+            return SKIPPED
+        else:
+            msg = f"Unrecognized status in meta.cal_step.{cal_step}."
+            raise ValueError(msg)
+        return getattr(datamodel[0].meta.cal_step, cal_step, NOT_SET)
```

Comment:
```suggestion
        msg = f"Unrecognized status in meta.cal_step.{cal_step}."
        raise ValueError(msg)
```
The `getattr(datamodel[0]...)` is unreachable.